File: examples/python/complex/advanced_patterns.py
"""
Advanced data processing pipeline demonstrating taint flow through
Python's most challenging syntax features. Each vulnerability chain
exercises a distinct syntax pattern that static analysis must handle.
"""

import os
import asyncio
import sqlite3
import subprocess
import json
import functools
from contextlib import contextmanager
from flask import Flask, request, jsonify, render_template_string


app = Flask(__name__)
db_path = os.environ.get("DATABASE_URL", "pipeline.db")


# ============================================================
# Helpers
# ============================================================

def get_connection():
    conn = sqlite3.connect(db_path)
    conn.row_factory = sqlite3.Row
    return conn


# ============================================================
# 1. Async/Await Chain
# VULN: request.args["cmd"] -> fetch_remote_config -> run_remote_task -> subprocess.run (command injection)
# ============================================================

async def fetch_remote_config(host):
    """Simulates fetching config from a remote host."""
    await asyncio.sleep(0)
    command = f"curl -s http://{host}/config"
    return command


async def run_remote_task(command):
    """Executes the fetched command string."""
    result = subprocess.run(command, shell=True, capture_output=True, text=True)
    return result.stdout


@app.route("/api/remote/config")
def trigger_remote_config():
    host = request.args["host"]
    loop = asyncio.new_event_loop()
    command = loop.run_until_complete(fetch_remote_config(host))
    output = loop.run_until_complete(run_remote_task(command))
    loop.close()
    return jsonify({"output": output})


# ============================================================
# 2. Generator / Yield
# VULN: request.args["filter"] -> stream_rows (generator) -> build_report_page -> render_template_string (XSS)
# ============================================================

def stream_rows(filter_expr):
    """Yields rows matching a user-supplied filter expression."""
    conn = get_connection()
    cursor = conn.cursor()
    cursor.execute("SELECT * FROM reports")
    for row in cursor.fetchall():
        if filter_expr in row["title"]:
            yield dict(row)
    conn.close()


def build_report_page(rows_iter):
    """Collects generator output into an HTML page."""
    parts = ["<h1>Report Results</h1><ul>"]
    for row in rows_iter:
        parts.append("<li>" + row["title"] + "</li>")
    parts.append("</ul>")
    return "".join(parts)


@app.route("/api/reports/stream")
def stream_reports():
    filter_text = request.args["filter"]
    rows = stream_rows(filter_text)
    html_page = build_report_page(rows)
    return render_template_string(html_page)


# ============================================================
# 3. Match/Case (Python 3.10+ Structural Pattern Matching)
# VULN: request.json["action"] -> dispatch_action (match/case) -> cursor.execute (SQL injection)
# ============================================================

def dispatch_action(action_payload):
    """Routes action through match/case, forwarding tainted data."""
    action_type = action_payload["type"]
    value = action_payload["value"]

    match action_type:
        case "query":
            return execute_custom_query(value)
        case "delete":
            return execute_custom_query(f"DELETE FROM records WHERE id = '{value}'")
        case "update":
            return execute_custom_query(f"UPDATE records SET status = 'done' WHERE name = '{value}'")
        case _:
            return {"error": "Unknown action"}


def execute_custom_query(sql):
    conn = get_connection()
    cursor = conn.cursor()
    cursor.execute(sql)
    conn.commit()
    results = cursor.fetchall()
    conn.close()
    return [dict(r) for r in results] if results else []


@app.route("/api/actions/dispatch", methods=["POST"])
def handle_action():
    payload = request.json
    result = dispatch_action(payload)
    return jsonify(result)


# ============================================================
# 4. Decorator Wrapping
# VULN: request.form["template"] -> render_with_log (decorator) -> inner render_template_string (template injection)
# ============================================================

def log_render(func):
    """Decorator that logs the rendered output and passes data through."""
    @functools.wraps(func)
    def wrapper(template_str, **kwargs):
        result = func(template_str, **kwargs)
        with open("/var/log/renders.log", "a") as f:
            f.write(f"Rendered: {template_str[:80]}\n")
        return result
    return wrapper


@log_render
def render_with_log(template_str, **kwargs):
    """Renders a user-supplied template after decorator wrapping."""
    return render_template_string(template_str, **kwargs)


@app.route("/api/templates/preview", methods=["POST"])
def preview_template():
    user_template = request.form["template"]
    username = request.form.get("username", "anonymous")
    output = render_with_log(user_template, user=username)
    return output


# ============================================================
# 5. List Comprehension -> SQL
# VULN: request.json["tags"] -> list comprehension build -> " OR ".join -> cursor.execute (SQL injection)
# ============================================================

def build_tag_query(tags):
    """Builds a SQL WHERE clause from a list of user-supplied tags."""
    conditions = ["tag = '" + t + "'" for t in tags]
    where_clause = " OR ".join(conditions)
    sql = "SELECT * FROM articles WHERE " + where_clause
    return sql


@app.route("/api/articles/by-tags", methods=["POST"])
def articles_by_tags():
    tags = request.json["tags"]
    sql = build_tag_query(tags)
    conn = get_connection()
    cursor = conn.cursor()
    cursor.execute(sql)
    results = [dict(row) for row in cursor.fetchall()]
    conn.close()
    return jsonify(results)


# ============================================================
# 6. Walrus Operator (:=)
# VULN: request.args["q"] -> walrus assignment -> format_search_header -> render_template_string (XSS)
# ============================================================

def format_search_header(query_text):
    """Formats query text into an HTML header string."""
    return f"<h2>Search results for: {query_text}</h2>"


@app.route("/api/search/walrus")
def walrus_search():
    if (query := request.args.get("q")) is not None:
        header = format_search_header(query)
        body = "<p>Processing your search...</p>"
        page = header + body
        return render_template_string(page)
    return jsonify({"error": "Missing query parameter"}), 400


# ============================================================
# 7. Closure Capture
# VULN: request.form["expr"] -> make_evaluator (closure) -> eval (code injection)
# ============================================================

def make_evaluator(expression):
    """Returns a closure that captures the user-supplied expression."""
    def evaluate():
        return eval(expression)
    return evaluate


@app.route("/api/calc", methods=["POST"])
def calculator():
    expr = request.form["expr"]
    evaluator = make_evaluator(expr)
    result = evaluator()
    return jsonify({"result": str(result)})


# ============================================================
# 8. Context Manager
# VULN: request.json["filename"] -> managed_file (context manager) -> open (path traversal)
# ============================================================

@contextmanager
def managed_file(path, mode="r"):
    """Context manager that opens a file from a user-supplied path."""
    f = open(path, mode)
    try:
        yield f
    finally:
        f.close()


@app.route("/api/files/read", methods=["POST"])
def read_user_file():
    filename = request.json["filename"]
    base_dir = "/var/app/data"
    full_path = os.path.join(base_dir, filename)
    with managed_file(full_path) as f:
        content = f.read()
    return jsonify({"content": content})


# ============================================================
# 9. *args Forwarding
# VULN: request.json["command"] -> run_task(*args) -> execute_shell(*args) -> os.system (command injection)
# ============================================================

def execute_shell(*args):
    """Executes a shell command assembled from *args."""
    cmd = " ".join(args)
    os.system(cmd)
    return {"status": "executed"}


def run_task(program, *args):
    """Forwards arguments through *args to the shell executor."""
    return execute_shell(program, *args)


@app.route("/api/tasks/run", methods=["POST"])
def run_user_task():
    program = request.json["command"]
    arguments = request.json.get("args", [])
    result = run_task(program, *arguments)
    return jsonify(result)


# ============================================================
# 10. F-string Multi-Hop
# VULN: request.args["name"] -> greet -> build_welcome_email -> os.system via sendmail (command injection)
# ============================================================

def greet(name):
    """Formats a greeting using the user-supplied name."""
    return f"Hello, {name}!"


def build_welcome_email(greeting, recipient):
    """Builds a sendmail command string from formatted greeting."""
    subject = f"Welcome - {greeting}"
    cmd = f"sendmail -t {recipient} -s '{subject}'"
    return cmd


@app.route("/api/welcome")
def welcome_user():
    name = request.args["name"]
    email = request.args["email"]
    greeting = greet(name)
    mail_cmd = build_welcome_email(greeting, email)
    os.system(mail_cmd)
    return jsonify({"status": "welcome email sent"})


# ============================================================
# 11. Higher-Order Functions (map/filter)
# VULN: request.json["ids"] -> map(build_clause, ...) -> join -> cursor.execute (SQL injection)
# ============================================================

def build_clause(record_id):
    """Builds a single SQL clause from a user-supplied ID."""
    return f"id = '{record_id}'"


@app.route("/api/records/bulk-lookup", methods=["POST"])
def bulk_lookup():
    ids = request.json["ids"]
    clauses = list(map(build_clause, ids))
    where = " OR ".join(clauses)
    sql = "SELECT * FROM records WHERE " + where
    conn = get_connection()
    cursor = conn.cursor()
    cursor.execute(sql)
    results = [dict(row) for row in cursor.fetchall()]
    conn.close()
    return jsonify(results)


# ============================================================
# 12. Generator Expression + filter (bonus higher-order variant)
# VULN: request.json["usernames"] -> filter -> generator expression -> cursor.execute (SQL injection)
# ============================================================

def is_valid_lookup(username):
    """Filter predicate -- allows any non-empty string through."""
    return len(username) > 0


def build_user_lookup(usernames):
    """Builds a SQL query from filtered usernames via generator expression."""
    valid = filter(is_valid_lookup, usernames)
    in_clause = ", ".join(f"'{u}'" for u in valid)
    return f"SELECT * FROM users WHERE username IN ({in_clause})"


@app.route("/api/users/bulk", methods=["POST"])
def bulk_user_lookup():
    usernames = request.json["usernames"]
    sql = build_user_lookup(usernames)
    conn = get_connection()
    cursor = conn.cursor()
    cursor.execute(sql)
    results = [dict(row) for row in cursor.fetchall()]
    conn.close()
    return jsonify(results)


# ============================================================
# 13. Class with __init__ Field Propagation + Method Chain
# VULN: request.json -> DataPipeline.__init__ -> transform -> persist -> cursor.execute (SQL injection)
# ============================================================

class DataPipeline:
    """Processes and stores data records with taint flowing through fields."""

    def __init__(self, source_name, raw_data):
        self.source_name = source_name
        self.raw_data = raw_data
        self.transformed = None

    def transform(self):
        """Transforms raw_data, preserving taint in self.transformed."""
        self.transformed = json.dumps(self.raw_data)
        return self

    def persist(self):
        """Stores transformed data via string-concatenated SQL."""
        conn = get_connection()
        cursor = conn.cursor()
        sql = "INSERT INTO pipeline_data (source, payload) VALUES ('" + self.source_name + "', '" + self.transformed + "')"
        cursor.execute(sql)
        conn.commit()
        conn.close()
        return {"status": "persisted"}


@app.route("/api/pipeline/ingest", methods=["POST"])
def ingest_pipeline():
    source = request.json["source"]
    data = request.json["data"]
    pipeline = DataPipeline(source, data)
    result = pipeline.transform().persist()
    return jsonify(result)


if __name__ == "__main__":
    app.run(debug=True, host="0.0.0.0", port=5001)

# ============================================================
# ADDITIONAL PATTERNS - Covering gaps found in flow audit
# ============================================================

# Comprehension with taint flow
def comprehension_flow(items):
    """Taint flows through list/dict comprehensions."""
    results = [eval(x) for x in items]  # list comp
    mapped = {k: eval(v) for k, v in items}  # dict comp
    first = next(x for x in items if x.startswith("cmd:"))  # genexpr
    return results

# Class inheritance chain with virtual dispatch
class BaseProcessor:
    def process(self, data):
        return self.transform(data)
    def transform(self, data):
        return data

class UnsafeProcessor(BaseProcessor):
    def transform(self, data):
        return eval(data)  # override reaches sink

class SafeProcessor(BaseProcessor):
    def transform(self, data):
        return str(data)  # safe override

# Decorator that wraps and passes through taint
def validate_input(func):
    @functools.wraps(func)
    def wrapper(*args, **kwargs):
        # Decorator passes taint through to wrapped function
        return func(*args, **kwargs)
    return wrapper

@validate_input
def execute_command(cmd):
    import os
    os.system(cmd)  # taint flows through decorator

# Exception flow
def exception_taint_flow(data):
    try:
        result = process_data(data)
    except ValueError as e:
        # Exception message may contain tainted data
        import logging
        logging.error(f"Failed: {e}")  # taint via exception
    finally:
        cleanup(data)

# Walrus operator in loop
def walrus_loop(stream):
    while (chunk := stream.read(1024)):
        eval(chunk)  # taint via walrus

# Dataclass with tainted fields
from dataclasses import dataclass

@dataclass
class UserRequest:
    username: str
    query: str

def handle_request(req: UserRequest):
    import os
    os.system(req.query)  # field access taint

# Lambda with taint flow
transform = lambda data: eval(data)  # taint through lambda

File: examples/go/complex/advanced_patterns.go
package main

import (
	"database/sql"
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"os/exec"
	"strings"
)

// ---------------------------------------------------------------------------
// advanced_patterns.go -- Go-specific syntax patterns for taint analysis
//
// Each vulnerability chain exercises a different Go language feature that
// taint must propagate through. Sources are always HTTP request values;
// sinks are db.Exec, exec.Command, os.Open, or fmt.Fprintf with user data.
// ---------------------------------------------------------------------------

// --- Helpers ---------------------------------------------------------------

type WorkItem struct {
	Payload string
	Done    bool
}

type Notifier interface {
	Send(msg string) error
}

type EmailNotifier struct {
	db *sql.DB
}

func (e *EmailNotifier) Send(msg string) error {
	// VULN 8b sink: interface dispatch lands here, msg is tainted
	// Chain: r.FormValue -> embedded field -> interface dispatch -> db.Exec
	_, err := e.db.Exec("INSERT INTO notifications (message) VALUES ('" + msg + "')")
	return err
}

type SlackNotifier struct{}

func (s *SlackNotifier) Send(msg string) error {
	fmt.Println("slack:", msg)
	return nil
}

// BaseService is embedded into AdvancedHandler to exercise struct embedding.
type BaseService struct {
	db *sql.DB
}

func (b *BaseService) RunRawSQL(query string) (*sql.Rows, error) {
	// VULN 8a sink: struct embedding, taint flows through embedded method
	return b.db.Query(query)
}

// AdvancedHandler embeds BaseService so callers can invoke RunRawSQL directly.
type AdvancedHandler struct {
	BaseService
	notifier Notifier
}

func NewAdvancedHandler(db *sql.DB) *AdvancedHandler {
	return &AdvancedHandler{
		BaseService: BaseService{db: db},
		notifier:    &EmailNotifier{db: db},
	}
}

// ---------------------------------------------------------------------------
// VULN 1 -- Channel send/receive flowing taint
//
// Chain: r.URL.Query().Get("q") -> ch <- q -> val := <-ch -> db.Exec(val)
// The tainted query parameter is sent into a channel, received on the other
// side, and used in an unsanitized SQL exec.
// ---------------------------------------------------------------------------

func (h *AdvancedHandler) ChannelFlow(w http.ResponseWriter, r *http.Request) {
	q := r.URL.Query().Get("q")
	ch := make(chan string, 1)
	ch <- q
	val := <-ch
	h.db.Exec("DELETE FROM cache WHERE key = '" + val + "'")
	fmt.Fprintf(w, "purged")
}

// ---------------------------------------------------------------------------
// VULN 2 -- Goroutine + closure capture
//
// Chain: r.FormValue("cmd") -> captured by goroutine closure -> exec.Command
// The closure captures `command` from the outer scope and executes it inside
// a goroutine. Taint must propagate through the closure capture.
// ---------------------------------------------------------------------------

func (h *AdvancedHandler) GoroutineClosure(w http.ResponseWriter, r *http.Request) {
	command := r.FormValue("cmd")
	done := make(chan []byte, 1)
	go func() {
		// VULN 2 sink: closure captures `command` from enclosing scope
		cmd := exec.Command("sh", "-c", command)
		out, _ := cmd.Output()
		done <- out
	}()
	result := <-done
	w.Write(result)
}

// ---------------------------------------------------------------------------
// VULN 3 -- Multiple return values (val, err)
//
// Chain: r.URL.Query().Get("path") -> fetchPath(path) returns (string, error)
//        -> content (first return value) -> os.Open(content)
// Taint flows through the first element of a multi-return.
// ---------------------------------------------------------------------------

func (h *AdvancedHandler) MultiReturn(w http.ResponseWriter, r *http.Request) {
	path := r.URL.Query().Get("path")
	resolved, err := resolvePath(path)
	if err != nil {
		http.Error(w, err.Error(), 400)
		return
	}
	// VULN 3 sink: resolved still carries taint from path
	f, err := os.Open(resolved)
	if err != nil {
		http.Error(w, "not found", 404)
		return
	}
	defer f.Close()
	fmt.Fprintf(w, "opened %s", resolved)
}

func resolvePath(input string) (string, error) {
	if input == "" {
		return "", fmt.Errorf("empty path")
	}
	// No sanitization -- passes taint through
	return "/data/" + input, nil
}

// ---------------------------------------------------------------------------
// VULN 4 -- Type switch (switch v := x.(type))
//
// Chain: json.NewDecoder(r.Body).Decode(&raw) -> type switch extracts string
//        -> db.Exec with that string
// Taint must propagate through the type assertion variable in each case arm.
// ---------------------------------------------------------------------------

func (h *AdvancedHandler) TypeSwitchFlow(w http.ResponseWriter, r *http.Request) {
	var raw interface{}
	json.NewDecoder(r.Body).Decode(&raw)

	var query string
	switch v := raw.(type) {
	case string:
		query = v
	case map[string]interface{}:
		if q, ok := v["query"]; ok {
			query = fmt.Sprintf("%v", q)
		}
	default:
		query = fmt.Sprintf("%v", v)
	}
	// VULN 4 sink: query derived from decoded body through type switch
	h.db.Exec("SELECT * FROM data WHERE filter = '" + query + "'")
	fmt.Fprintf(w, "ok")
}

// ---------------------------------------------------------------------------
// VULN 5 -- Select statement
//
// Chain: r.FormValue("host") -> sent into one of two channels -> select
//        receives from whichever is ready -> exec.Command("ping", target)
// Taint propagates through a select receiving from multiple channels.
// ---------------------------------------------------------------------------

func (h *AdvancedHandler) SelectFlow(w http.ResponseWriter, r *http.Request) {
	host := r.FormValue("host")
	primary := make(chan string, 1)
	fallback := make(chan string, 1)
	primary <- host

	var target string
	select {
	case target = <-primary:
		// received from primary
	case target = <-fallback:
		// received from fallback
	}
	// VULN 5 sink: target carries taint from host
	cmd := exec.Command("ping", "-c", "1", target)
	out, _ := cmd.Output()
	w.Write(out)
}

// ---------------------------------------------------------------------------
// VULN 6 -- Named return values
//
// Chain: r.URL.Query().Get("filter") -> buildNamedFilter sets named return
//        -> caller uses returned value in db.Exec
// Taint flows through a named return variable that is set in the function
// body and returned implicitly via bare `return`.
// ---------------------------------------------------------------------------

func (h *AdvancedHandler) NamedReturnFlow(w http.ResponseWriter, r *http.Request) {
	filter := r.URL.Query().Get("filter")
	clause := buildNamedFilter(filter)
	// VULN 6 sink: clause carries taint from filter
	h.db.Exec("SELECT * FROM events WHERE " + clause)
	fmt.Fprintf(w, "done")
}

func buildNamedFilter(input string) (result string) {
	result = "status = 'active' AND name LIKE '%" + input + "%'"
	return
}

// ---------------------------------------------------------------------------
// VULN 7 -- Defer + closure
//
// Chain: r.FormValue("tag") -> deferred closure captures `tag` -> at function
//        exit the closure runs db.Exec with tainted tag
// Taint must propagate through a deferred closure's captured variables.
// ---------------------------------------------------------------------------

func (h *AdvancedHandler) DeferClosure(w http.ResponseWriter, r *http.Request) {
	tag := r.FormValue("tag")
	defer func() {
		// VULN 7 sink: deferred closure captures `tag`, executes at exit
		h.db.Exec("INSERT INTO audit_log (tag) VALUES ('" + tag + "')")
	}()
	fmt.Fprintf(w, "processing tag=%s", tag)
}

// ---------------------------------------------------------------------------
// VULN 8a -- Struct embedding
//
// Chain: r.URL.Query().Get("table") -> direct call to embedded method
//        h.RunRawSQL(query) which is defined on BaseService
// Taint flows through an embedded struct's method.
// ---------------------------------------------------------------------------

func (h *AdvancedHandler) EmbeddedMethodCall(w http.ResponseWriter, r *http.Request) {
	table := r.URL.Query().Get("table")
	query := "SELECT * FROM " + table
	// RunRawSQL is promoted from the embedded BaseService
	rows, err := h.RunRawSQL(query)
	if err != nil {
		http.Error(w, err.Error(), 500)
		return
	}
	defer rows.Close()
	fmt.Fprintf(w, "queried %s", table)
}

// ---------------------------------------------------------------------------
// VULN 8b -- Interface dispatch
//
// Chain: r.FormValue("message") -> h.notifier.Send(msg) -> dispatches to
//        EmailNotifier.Send -> db.Exec with tainted msg
// Taint flows through an interface method call to the concrete implementation.
// ---------------------------------------------------------------------------

func (h *AdvancedHandler) InterfaceDispatch(w http.ResponseWriter, r *http.Request) {
	msg := r.FormValue("message")
	err := h.notifier.Send(msg)
	if err != nil {
		http.Error(w, err.Error(), 500)
		return
	}
	fmt.Fprintf(w, "notified")
}

// ---------------------------------------------------------------------------
// VULN 9 -- strings.Builder chain
//
// Chain: r.URL.Query().Get("name") -> strings.Builder.WriteString (x3) ->
//        builder.String() -> db.Exec
// Taint propagates through a Builder's accumulated writes to String().
// ---------------------------------------------------------------------------

func (h *AdvancedHandler) BuilderChain(w http.ResponseWriter, r *http.Request) {
	name := r.URL.Query().Get("name")
	var b strings.Builder
	b.WriteString("UPDATE users SET last_login = NOW() WHERE name = '")
	b.WriteString(name)
	b.WriteString("'")
	stmt := b.String()
	// VULN 9 sink: stmt built with tainted name via strings.Builder
	h.db.Exec(stmt)
	fmt.Fprintf(w, "updated")
}

// ---------------------------------------------------------------------------
// VULN 10 -- Variadic function (...string)
//
// Chain: r.FormValue("a"), r.FormValue("b"), r.FormValue("c") -> joinParts
//        (variadic) -> concatenated -> db.Exec
// Taint must flow through variadic parameters into the joined result.
// ---------------------------------------------------------------------------

func (h *AdvancedHandler) VariadicFlow(w http.ResponseWriter, r *http.Request) {
	a := r.FormValue("a")
	b := r.FormValue("b")
	c := r.FormValue("c")
	clause := joinParts(" AND ", a, b, c)
	// VULN 10 sink: clause built from tainted variadic args
	h.db.Exec("SELECT * FROM records WHERE " + clause)
	fmt.Fprintf(w, "ok")
}

func joinParts(sep string, parts ...string) string {
	return strings.Join(parts, sep)
}

// ---------------------------------------------------------------------------
// VULN 11 -- If-init statement (if val := expr; cond)
//
// Chain: r.URL.Query().Get("file") -> if-init assignment with condition ->
//        os.Open(validated) inside the if body
// Taint propagates through the short variable in an if-init statement.
// ---------------------------------------------------------------------------

func (h *AdvancedHandler) IfInitFlow(w http.ResponseWriter, r *http.Request) {
	file := r.URL.Query().Get("file")
	if validated := validateFilename(file); validated != "" {
		// VULN 11 sink: validated still tainted (validateFilename is a no-op)
		f, err := os.Open(validated)
		if err != nil {
			http.Error(w, "not found", 404)
			return
		}
		defer f.Close()
		fmt.Fprintf(w, "opened %s", validated)
	}
}

func validateFilename(name string) string {
	// Bug: no actual sanitization
	if name == "" {
		return ""
	}
	return name
}

// ---------------------------------------------------------------------------
// VULN 12 -- Goroutine channel pipeline (multi-stage)
//
// Chain: r.FormValue("expr") -> stage1 channel -> transform -> stage2 channel
//        -> db.Exec. A two-stage pipeline where taint must cross two channel
//        boundaries.
// ---------------------------------------------------------------------------

func (h *AdvancedHandler) PipelineFlow(w http.ResponseWriter, r *http.Request) {
	expr := r.FormValue("expr")

	stage1 := make(chan string, 1)
	stage2 := make(chan string, 1)

	// Stage 1: prefix the expression
	go func() {
		val := <-stage1
		stage2 <- "computed_" + val
	}()

	stage1 <- expr
	result := <-stage2
	// VULN 12 sink: result carries taint from expr through two channels
	h.db.Exec("INSERT INTO results (expr) VALUES ('" + result + "')")
	fmt.Fprintf(w, "computed")
}

// ---------------------------------------------------------------------------
// VULN 13 -- Range over slice with tainted elements
//
// Chain: json.NewDecoder(r.Body).Decode(&items) -> range loop -> each item
//        used in db.Exec
// Taint propagates through range iteration over a decoded slice.
// ---------------------------------------------------------------------------

func (h *AdvancedHandler) RangeFlow(w http.ResponseWriter, r *http.Request) {
	var items []string
	json.NewDecoder(r.Body).Decode(&items)
	for _, item := range items {
		// VULN 13 sink: each item is tainted from the decoded body
		h.db.Exec("INSERT INTO batch (value) VALUES ('" + item + "')")
	}
	fmt.Fprintf(w, "inserted %d items", len(items))
}

// ---------------------------------------------------------------------------
// VULN 14 -- Map lookup with tainted value
//
// Chain: r.URL.Query().Get("key") -> map[key] retrieves a previously stored
//        tainted value -> db.Exec. Taint stored in a map and retrieved later.
// ---------------------------------------------------------------------------

func (h *AdvancedHandler) MapLookupFlow(w http.ResponseWriter, r *http.Request) {
	data := make(map[string]string)
	data["user"] = r.URL.Query().Get("user")
	data["action"] = r.URL.Query().Get("action")

	userVal := data["user"]
	// VULN 14 sink: userVal retrieved from map still carries taint
	h.db.Exec("INSERT INTO activity (user_name) VALUES ('" + userVal + "')")
	fmt.Fprintf(w, "logged %s", userVal)
}

// ---------------------------------------------------------------------------
// VULN 15 -- Struct field assignment and read-back
//
// Chain: r.FormValue("payload") -> WorkItem{Payload: payload} -> item.Payload
//        -> exec.Command. Taint flows through struct field write then read.
// ---------------------------------------------------------------------------

func (h *AdvancedHandler) StructFieldFlow(w http.ResponseWriter, r *http.Request) {
	payload := r.FormValue("payload")
	item := WorkItem{Payload: payload}
	// VULN 15 sink: item.Payload carries taint from FormValue
	cmd := exec.Command("sh", "-c", item.Payload)
	out, _ := cmd.Output()
	w.Write(out)
}

// ============================================================
// ADDITIONAL PATTERNS - Error wrapping, context propagation
// ============================================================

// Error wrapping with taint
func errorWrapFlow(input string) error {
	result, err := processData(input)
	if err != nil {
		return fmt.Errorf("failed to process %s: %w", input, err)
	}
	exec.Command(result).Run()
	return nil
}

// Context value propagation
func contextFlow(ctx context.Context) {
	cmd := ctx.Value("user_command").(string)
	exec.Command(cmd).Run()  // taint from context value
}

// Slice append propagation
func sliceFlow(inputs []string) {
	var commands []string
	for _, input := range inputs {
		commands = append(commands, input)
	}
	for _, cmd := range commands {
		exec.Command(cmd).Run()  // taint through slice
	}
}

// Map propagation
func mapFlow(input string) {
	config := make(map[string]string)
	config["cmd"] = input
	exec.Command(config["cmd"]).Run()  // taint through map
}
